File: lab3/egg.py
from OpenGL.GL import *

import utils
import numpy as np

dimensions = utils.Coord(600, 600)
bounds = utils.Coord(7.5, 7.5)
viewport = utils.Viewport(bounds, 8, 8)
N = 50
vertices = np.zeros((N, N, 3))


def startup():
    glEnable(GL_DEPTH_TEST)
    glClearColor(0.5, 0.5, 0.5, 0.1)
    viewport(None, dimensions)
    make_egg_vertices()


def shutdown():
    pass


def egg_lines(coord: utils.Coord, color):
    glBegin(GL_LINES)
    glColor3ubv(color)

    glVertex3fv(vertices[coord.x][coord.y])
    glVertex3fv(vertices[coord.x + 1][coord.y])
    glEnd()

    glBegin(GL_LINES)
    glColor3ubv(color)

    glVertex3fv(vertices[coord.x][coord.y])
    glVertex3fv(vertices[coord.x][coord.y + 1])
    glEnd()


def egg_triangles(coord: utils.Coord, color):
    glBegin(GL_TRIANGLES)
    glColor3ubv(color)
    glVertex3fv(vertices[coord.x][coord.y])
    glVertex3fv(vertices[coord.x + 1][coord.y])
    glVertex3fv(vertices[coord.x][coord.y + 1])
    glEnd()

    glBegin(GL_TRIANGLES)
    glColor3ubv(color)
    glVertex3fv(vertices[coord.x + 1][coord.y + 1])
    glVertex3fv(vertices[coord.x + 1][coord.y])
    glVertex3fv(vertices[coord.x][coord.y + 1])
    glEnd()


def egg_strips(i, color):
    glBegin(GL_TRIANGLE_STRIP)
    glColor3ubv(color)
    for j in range(N - 1):
        glVertex3fv(vertices[i][j])
        glVertex3fv(vertices[i + 1][j + 1])
    glVertex3fv(vertices[-1][-1])
    glVertex3fv(vertices[0][0])
    glEnd()


def make_egg_vertices():
    for i in range(N):
        for j in range(N):
            u = i / (N - 1)
            v = j / (N - 1)
            vertices[i][j][0] = (-90 * u ** 5 + 225 * u ** 4 - 270 * u ** 3 + 180 * u ** 2 - 45 * u) * np.cos(np.pi * v)
            vertices[i][j][1] = 160 * u ** 4 - 320 * u ** 3 + 160 * u ** 2 - 5
            vertices[i][j][2] = (-90 * u ** 5 + 225 * u ** 4 - 270 * u ** 3 + 180 * u ** 2 - 45 * u) * np.sin(np.pi * v)


def render(time):
    glClear(GL_DEPTH_BUFFER_BIT | GL_COLOR_BUFFER_BIT)
    glLoadIdentity()

    utils.spin(time * 180 / np.pi)

    utils.axes(6, 6, 6)

    for i in range(N - 1):
        egg_strips(vertices, i, [255, 172, 28])

    glFlush()


def main():
    utils.default_init(dimensions=dimensions,
                       startup=startup,
                       render=render,
                       shutdown=shutdown,
                       viewport=viewport)


if __name__ == '__main__':
    main()

File: lab3/Egg.cpp
#if defined(_WIN32)
#if !defined(NOMINMAX)
#define NOMINMAX
#endif
#include <Windows.h>
#endif
#include <GLFW/glfw3.h>

#include <cmath>
#include <numbers>

#include "Egg.h"

namespace gk::lab3
{

auto Egg::run(Vector2i dimensions, uint32_t resolution, bool vsync) -> void
{
    this->dimensions = dimensions;
    vertices = makeEggVertices(resolution);

    utils::configureViewport(bounds, nearDepth, farDepth);
    utils::defaultInit(dimensions,
                       [this] { startup(); },
                       [this] (auto) { render(); },
                       [this] { shutdown(); },
                       utils::Config::Dimensions::_2D,
                       vsync);
}

auto Egg::startup() const -> void
{
    glEnable(GL_DEPTH_TEST);
    glClearColor(0.5f, 0.5f, 0.5f, 0.1f);
    utils::viewport(nullptr, dimensions.x, dimensions.y);
}

auto Egg::shutdown() -> void
{

}

auto Egg::render() -> void
{
    glClear(GL_DEPTH_BUFFER_BIT | GL_COLOR_BUFFER_BIT);
    glLoadIdentity();

    utils::spin(static_cast<float>(glfwGetTime() * 180.0 / std::numbers::pi));
    utils::axes(6, 6, 6);
    makeEgg();
    glFlush();
}

auto Egg::makeEgg() -> void
{
    glBegin(GL_TRIANGLE_STRIP);
    for (auto i = size_t {}; i < vertices.size() - 1; i++)
    {
        for (auto j = size_t {}; j < vertices.size(); j++)
        {
            glColor3fv(vertices[i][j].second.toBytes());
            glVertex3fv(vertices[i][j].first.toBytes());

            glColor3fv(vertices[i + 1][j].second.toBytes());
            glVertex3fv(vertices[i + 1][j].first.toBytes());
        }
    }
    glEnd();
}

auto Egg::makeEggVertices(size_t N) -> std::vector<std::vector<std::pair<Vector3f, Color3f>>>
{
    static auto vertices = std::vector(N, std::vector(N, std::pair<Vector3f, Color3f>{}));

    for (auto i = size_t {}; i < N; i++)
    {
        for (auto j = size_t {}; j < N; j++)
        {
            const auto u = static_cast<float>(i) / static_cast<float>(N - 1);
            const auto v = static_cast<float>(j) / static_cast<float>(N - 1);
            vertices[i][j].first.x = static_cast<float>(
                (-90. * std::pow(u, 5) + 225. * std::pow(u, 4) - 270. * std::pow(u, 3) + 180. * std::pow(u, 2) -
                 45. * u) * std::cos(std::numbers::pi * v));
            vertices[i][j].first.y = static_cast<float>(160. * std::pow(u, 4) - 320. * std::pow(u, 3) +
                                                   160. * std::pow(u, 2) - 5.);
            vertices[i][j].first.z = static_cast<float>(
                (-90. * std::pow(u, 5) + 225. * std::pow(u, 4) - 270. * std::pow(u, 3) + 180. * std::pow(u, 2) -
                 45. * u) * std::sin(std::numbers::pi * v));

            const auto colorVal = std::lerp(0.f, 1.f, u);
            vertices[i][j].second = {0.5f, colorVal, 0.7f};
        }
    }
    return vertices;
}


}

